### DISCUSSION TITLE
Failed to build a native library through cargo

### DISCUSSION THREAD

**Discussion by jblindsay** [messense/maturin-action#78] [Q&A] [answered]
Hello, I am trying to use the maturin-actions command and am having my builds fail with:

```
run pkg_config fail: "`\"pkg-config\" \"--libs\" \"--cflags\" \"openssl\"` did not exit successfully: exit status: 1\nerror: could not find system library 'openssl' required by the 'openssl-sys' crate\n\n--- stderr\nPackage openssl was not found in the pkg-config search path.\nPerhaps you should add the directory containing `openssl.pc'\nto the PKG_CONFIG_PATH environment variable\nNo package 'openssl' found\n"

--- stderr
  thread 'main' panicked at '

  Could not find directory of OpenSSL installation, and this `-sys` crate cannot
  proceed without this knowledge. If OpenSSL is installed and this crate had
  trouble finding it,  you can set the `OPENSSL_DIR` environment variable for the
  compilation process.

  Make sure you also have the development packages of openssl installed.
  For example, `libssl-dev` on Ubuntu or `openssl-devel` on Fedora.

  If you're in a situation where you think the directory *should* be found
  automatically, please open a bug at https://github.com/sfackler/rust-openssl
  and include information about your system as well as this message.

  $HOST = x86_64-unknown-linux-gnu
  $TARGET = x86_64-unknown-linux-gnu
  openssl-sys = 0.9.75
```

One of my dependent crates requires openssl. Have you encountered this issue before and do you know how to resolve it? Thank you very much?

**Comment by messense** [comment, marked as answer]
Static link `openssl` is the easiest solution, [activate the `vendered` feature](https://docs.rs/openssl/latest/openssl/#vendored) of `openssl` crate should do.

**Comment by jblindsay** [comment]
Thanks for the quick response. Unfortunately, I'm not directly depending on openssl. I think it's getting brought in as a dependency of a dependency, possibly Tokio, but I really don't know for sure.

**Reply by messense** [reply]
Cargo does [feature unification](https://doc.rust-lang.org/cargo/reference/features.html#feature-unification), so you can force a static openssl build by adding it to your dependencies.

```toml
[dependencies]
openssl = { version = "0.10", features = ["vendored"] }
```

**Reply by jblindsay** [reply]
genius!

**Reply by jblindsay** [reply]
I just wanted to let you know that this worked perfectly for me. Now I have another question, and I do apologize if this is really basic stuff but this is all new to me. I'd like to publish the resulting artifact to the PyPi repo. Typically I have been just using `maturin publish` at the command line on my Mac and Windows machines to publish the artifacts, using my -u and -p flags to provide username and password info. So here's my question, how is that completed with the Github action? I have seen this in the publish part of the action, `MATURIN_PASSWORD: ${{ secrets.PYPI_TOKEN }}` and I assume that it is retrieving this same info from some 'secrets' environmental setting. But how do I set that up to work with my profile? Or alternatively, and here I realize that I must sound like a fool to you so again I apologize, but is there a way that I can manually publish the wheel that the Github action produces? So sorry to keep bothering you...

**Reply by jblindsay** [reply]
I figured out how to do it with Twine. Thanks for all of your help today. It is much appreciated.
